### ISSUE TITLE
Add support for MSDTC on Windows Containers

### ISSUE THREAD

**Issue by vrapolinario** [microsoft/Windows-Containers#24]
Request from User Voice: https://windowsserver.uservoice.com/forums/304624-containers/suggestions/30982237-support-for-msdtc

We have legacy .NET based business applications making extensive use of MSDTC. It will be good to have support for this role in containers to enable Lift & Shift scenarios.

**Comment by vrapolinario** [comment]
The team is now investigating this functionality. We'd love to have customers help us validate the operation of MSDTC on Windows Containers. Please let us know if you'd like to support this.

**Comment by vrapolinario** [comment]
Sorry for not making any comments on this.
This is still in progress but we don't have any updates as of now.
More to come soon.

**Comment by vrapolinario** [comment]
We're currently working on documentation about the supported scenarios for MSDTC on Windows containers and we'll publish this soon. In the meantime, I'm curious to know if we have customers still blocked on this.

**Comment by mr-mariano-be** [comment]
Hi, at our organization we are currently analysing moving to a container infrastructure many production .NET4.x web applications running in different versions of IIS that currently uses MSDTC extensively when connecting to WCF services. Non MSDTC support would be blocking, but it is still work in progress and delayed due to COVID so we haven't get to an functional PoC or lab environment.

Thanks!

**Comment by vrapolinario** [comment]
To clarify, MSDTC does work on Windows Containers and is a supported scenario. Our team was able to validate its usage and confirmed that even with gMSA it is possible to run MSDTC. We do know if issues related to networking when using MSDTC on K8s and that is out of scope for now. The documentation on our page should be out soon.

**Comment by jasper22** [comment]
@vrapolinario  Any updates ? My organization currently stuck with this problem while trying to contenerize .NET Framework 4.5 application and SQL Server in different containers.  .NET always throws an exception that involving MSDTC..... We will be very happy just to look at the draft of your documentations - it would be **very** helpfull

**Comment by ohault** [comment]
Support of MSDTC is just part of a broader use-case. There is the same exact question about SMB service and many others NT services. It's all about muti-tenancy, and two paths could be followed:
1) native multi-tenancy of the NT service implementation itself, but there is still the problem with single-tenancy aspect of LSASS
2) multi-tenancy by containers (thanks to gMSA) where current single-tenancy NT service implementation could be "multiplexed"
Based on specific needs the isolation level (process, VBS, ...) should only be a configuration parameter.
One of the key point it about scalability and avoiding any boatware approach using plain VMs.
The Component-Based Servicing approach could help a lot to build custom based images with only the minimum required set of files based on the NT service you need to run in a container.
I would be pleased to understand the path followed to implement Azure Files service.

**Comment by derek987654** [comment]
> We're currently working on documentation about the supported scenarios for MSDTC on Windows containers and we'll publish this soon. In the meantime, I'm curious to know if we have customers still blocked on this.

We are blocked by this - trying to avoid rewriting a very large codebase of C++ critical business logic.  Rather, we want to move to .NET Core, but it will take time.  Since we can't do everything at once, the ability to run MSDTC in a windows container would be very valuable to us.

Any idea when the documentation/updates might be available?

**Comment by ghost** [comment]
This issue has been open for 90 days with no updates.
@vrapolinario, please provide an update or close this issue.

**Comment by weijuans-msft** [comment]
@derek987654 @ohault @jasper22 @mr-mariano-be  We are taking another look at this ask here. Making MSDTC working in containers is not trivial work. Hence I wanted to see if this is still needed. Are you blocked on moving to Windows containers with this? or have you found alternative options? Either way, would love to hear from the community so we can decide. Thanks!

**Comment by ohault** [comment]
For sure, it's not a trivial work but it is a very good use-case of how Windows Containers design and implementation should continue to evolve. In fact, we have the same question for all others Windows Server services. It's strategic as a paradigm shift towards running them into Kubernetes (let's say also in Azure Stack HCI).

**Comment by AF250329** [comment]
> We are taking another look at this ask here. 

We found a technical solution to this problem and were able to get it running, but everyone here is convinced that SQL Server with MSDTC is a major issue on containers. The next version of our application will rewrite this section so that it can be used with any SQL server without the need for MSDTC.

**Comment by weijuans-msft** [comment]
@AF250329   Wow, that sounds amazing! I would love to learn how you guys did it, and what you plan to do in your next version.

**Comment by AF250329** [comment]
> @AF250329 Wow, that sounds amazing! 

It a little bit hackish solution: We put whole SQL Server into the same container where .NET Framework application running.  As we discovered, opening a connection string to 'localhost' disables the MSDTC service. This configuration adds an extra 6-8 Gb of space to the container, but at least we're using MSDTC.

We are attempting to avoid triggering MSDTC in any situation in the next version. Some even recommend switching to Postrgress or MySql, etc...

**Comment by derek987654** [comment]
> @derek987654 @ohault @jasper22 @mr-mariano-be We are taking another look at this ask here. 


This is definitely a blocking issue for us as we require MSDTC. However, I think more of a concern to us is what appears to be a lack of support for MSDTC within .NET6.0.  So even if you solve this MSDTC/Container issue for us at the moment, lacking direct support for MSDTC in .NET6.0 means that we are still going to have to address the fundamental concern.  I would love to understand how people do distributed transactions including both a database table and a message queue without MSDTC (or, of course, database tables residing in separate databases).  My reading tells me that utilizing an Outbox Pattern seems to be the recommended approach, but that is clearly fundamentally different than an actual distributed transaction implementation.  

So yes it is blocking our usage of Windows Containers, but since MSDTC does not appear to be supported in .NET6.0, your fixing this will not help us in the long term.

Thank you for reaching out.

**Comment by mr-mariano-be** [comment]
For us is a blocking issue. In our case, we are not trying to attach only MSSQL. We haven't analysed others alternatives for now as the impact would be larger. Thanks!

**Comment by weijuans-msft** [comment]
+ @richlander from .NET team.

@derek987654  Interesting to see you are using .NET 6 on Windows containers. Would love to learn more the use cases/scenarios. Feel free to email us at win-containers@microsoft.com.

@mr-mariano-be and all: to be fully transparent, I need data to make a business justification. Please do send us emails privately and share how big is the blocker, e.g., how many apps/VMs are impacted. Or share here if you are ok.

**Comment by ajcvickers** [comment]
Distributed transactions are not supported on .NET 6. We are investigating adding Windows-only support for use of DTC through System.Transactions for .NET 7. This is tracked here: https://github.com/dotnet/runtime/issues/715#issuecomment-997271773

**Comment by pinzhongwang** [comment]
Just pray to be able to support DTC as soon as possible, DTC is the core of our microservices architecture, nearly 20 years of architectural accumulation, really unwilling to switch to Java, the amount of code involved is too large. The current server is still based on .net4.6.1, although the web client has been upgraded to net6.0 Blazor, has been looking forward to the addition of DTC and the release of Maui on the App side. For SQLServer to support more than 50 billion of active data, the overall architecture still takes a lot of thought and does not want to give up. Although the current microservice side has added cross-platform call support for Java and JS to prepare for switching, it is still possible to switch to Java or not to switch to Java.

**Comment by mthalman** [comment]
> The current server is still based on .net4.6.1

@pinzhongwang - Be aware that [support for .NET Framework 4.6.1 ended](https://docs.microsoft.com/en-us/lifecycle/products/microsoft-net-framework) on April 26, 2022. It's highly recommended you upgrade to a supported version.

**Comment by pinzhongwang** [comment]
> 

Thank you to point out that the current server-side system is running very stable, just wait for the new DTC support, as long as .net70 supports DTC, our middleware can be upgraded within 1 week, the business layer code does not need to be modified, recompile it can be. In addition to the basic modules related to DTC-related database access, other code has supported .net6.0, including Maui Blazor is also working fine, and the client has begun to switch to the module with Blazor as the core. It is still very gratifying to see that DTC support is already on the agenda, and it can be expected.

**Comment by cxx7532706** [comment]
It's also a blocking issue for us at this moment, any updates on this?

**Comment by akarshm** [comment]
The details for support are captured [here](https://learn.microsoft.com/en-us/virtualization/windowscontainers/quick-start/lift-shift-to-containers#msdtc).

**Comment by tom-haines** [comment]
It's also a blocking issue for us at this moment, any updates on this?

**Comment by pinzhongwang** [comment]
the server side  back to net4.6.2/net4.8.1, the Web client use net 8.0, it is still very stable to use, if  use the docker, only use the windows docker, the other has no impact. Stability comes first




***@***.***
 
From: Tom Haines
Date: 2024-09-04 17:36
To: microsoft/Windows-Containers
CC: pinzhongwang; Mention
Subject: Re: [microsoft/Windows-Containers] Add support for MSDTC on Windows Containers (#24)
It's also a blocking issue for us at this moment, any updates on this?
